File: packages/python-sdk/e2b/templates/data_analysis.py
import logging
import time
from typing import Optional, Callable, Any, Tuple, List, Union
from e2b.sandbox.process import ProcessMessage

from pydantic import BaseModel, PrivateAttr

from e2b import EnvVars, Sandbox
from e2b.constants import TIMEOUT

logger = logging.getLogger(__name__)


class Artifact(BaseModel):
    name: str
    _sandbox: Sandbox = PrivateAttr()

    def __init__(self, sandbox: Sandbox, **data: Any):
        super().__init__(**data)
        self._sandbox = sandbox

    def __hash__(self):
        return hash(self.name)

    def download(self) -> bytes:
        return self._sandbox.download_file(self.name)


class DataAnalysis(Sandbox):
    env_id = "Python3-DataAnalysis"

    def __init__(
        self,
        api_key: Optional[str] = None,
        cwd: Optional[str] = None,
        env_vars: Optional[EnvVars] = None,
        timeout: Optional[float] = TIMEOUT,
        on_stdout: Optional[Callable[[ProcessMessage], Any]] = None,
        on_stderr: Optional[Callable[[ProcessMessage], Any]] = None,
        on_artifact: Optional[Callable[[Artifact], Any]] = None,
        on_exit: Optional[Callable[[int], Any]] = None,
        **kwargs,
    ):
        self.on_artifact = on_artifact
        super().__init__(
            id=self.env_id,
            api_key=api_key,
            cwd=cwd,
            env_vars=env_vars,
            timeout=timeout,
            on_stdout=on_stdout,
            on_stderr=on_stderr,
            on_exit=on_exit,
            **kwargs,
        )

    def run_python(
        self,
        code: str,
        on_stdout: Optional[Callable[[ProcessMessage], Any]] = None,
        on_stderr: Optional[Callable[[ProcessMessage], Any]] = None,
        on_artifact: Optional[Callable[[Artifact], Any]] = None,
        on_exit: Optional[Callable[[int], Any]] = None,
        env_vars: Optional[EnvVars] = None,
        cwd: str = "",
        process_id: Optional[str] = None,
        timeout: Optional[float] = TIMEOUT,
    ) -> Tuple[str, str, List[Artifact]]:
        artifacts = set()

        def register_artifacts(event: Any) -> None:
            on_artifact_func = on_artifact or self.on_artifact
            if event.operation == "Create":
                artifact = Artifact(name=event.path, sandbox=self)
                artifacts.add(artifact)
                if on_artifact_func:
                    try:
                        on_artifact_func(artifact)
                    except Exception as e:
                        logger.error("Failed to process artifact", exc_info=e)

        watcher = self.filesystem.watch_dir("/home/user/artifacts")
        watcher.add_event_listener(register_artifacts)
        watcher.start()

        epoch_time = time.time()
        codefile_path = f"/tmp/main-{epoch_time}.py"
        self.filesystem.write(codefile_path, code)

        process = self.process.start(
            f"python {codefile_path}",
            on_stdout=on_stdout,
            on_stderr=on_stderr,
            on_exit=on_exit,
            env_vars=env_vars,
            cwd=cwd,
            process_id=process_id,
            timeout=timeout,
        )
        process.wait()

        watcher.stop()

        return process.output.stdout, process.output.stderr, list(artifacts)

    def _install_packages(
        self,
        command: str,
        package_names: Union[str, List[str]],
        timeout: Optional[float] = TIMEOUT,
    ) -> None:
        if isinstance(package_names, list):
            package_names = " ".join(package_names)

        package_names = package_names.strip()
        if not package_names:
            return

        process = self.process.start(f"{command} {package_names}", timeout=timeout)
        process.wait()

        if process.exit_code != 0:
            raise Exception(
                f"Failed to install package {package_names}: {process.output.stderr}"
            )

    def install_python_packages(
        self, package_names: Union[str, List[str]], timeout: Optional[float] = TIMEOUT
    ) -> None:
        self._install_packages("pip install", package_names, timeout=timeout)

    def install_system_packages(
        self, package_names: Union[str, List[str]], timeout: Optional[float] = TIMEOUT
    ) -> None:
        self._install_packages(
            "sudo apt-get install -y", package_names, timeout=timeout
        )


CodeInterpreter = DataAnalysis

File: packages/js-sdk/src/templates/dataAnalysis.ts
import { Sandbox, SandboxOpts } from '../sandbox'
import type { ProcessOpts } from '../sandbox/process'
import { FilesystemEvent, FilesystemOperation } from '../sandbox/filesystemWatcher'

export class Artifact<S extends DataAnalysis> {
  readonly path: string
  readonly _sandbox: S

  constructor(path: string, sandbox: S) {
    this.path = path
    this._sandbox = sandbox
  }
}

export interface RunPythonOpts<S extends DataAnalysis> extends Omit<ProcessOpts, 'cmd'> {
  onArtifact?: (artifact: Artifact<S>) => Promise<void> | void;
}

export class DataAnalysis extends Sandbox {
  private static template = 'Python3-DataAnalysis'

  /**
   * Use `DataAnalysis.create()` instead.
   * 
   * @hidden
   * @hide
   * @internal
   * @access protected
   */
  constructor(opts: SandboxOpts) {
    super({ template: opts.template || DataAnalysis.template, ...opts })
  }

  /**
   * Creates a new Sandbox from the template.
   * @returns New Sandbox
   */
  static override async create(): Promise<DataAnalysis>;
  /**
   * Creates a new Sandbox from the specified options.
   * @param opts Sandbox options
   * @returns New Sandbox
   */
  static override async create(opts: SandboxOpts): Promise<DataAnalysis>;
  static override async create(opts?: SandboxOpts) {
    const sandbox = new DataAnalysis({ ...opts ? opts : {} })
    await sandbox._open({ timeout: opts?.timeout })

    return sandbox
  }

  async runPython(code: string, opts: RunPythonOpts<this> = {}): Promise<{
    stdout: string;
    stderr: string;
    artifacts: Artifact<DataAnalysis>[];
  }> {
    const artifacts: string[] = []

    const registerArtifacts = async (event: FilesystemEvent) => {
      if (event.operation === FilesystemOperation.Create) {
        const artifact = new Artifact<this>(event.path, this)
        artifacts.push(event.path)
        await opts.onArtifact?.(artifact)
      }
    }

    const watcher = this.filesystem.watchDir('/home/user/artifacts')
    watcher.addEventListener(registerArtifacts)
    await watcher.start()

    const currentEpoch = new Date().getTime()
    const codefilePath = `/tmp/main-${currentEpoch}.py`
    await this.filesystem.write(codefilePath, code)
    const output = await this.process.startAndWait({
      cmd: `python ${codefilePath}`,
      ...opts,
    })

    await watcher.stop()

    return {
      stdout: output.stdout,
      stderr: output.stderr,
      artifacts: artifacts.map((artifact) => new Artifact<this>(artifact, this)),
    }
  }

  async installPythonPackages(packageNames: string | string[]) {
    await this.installPackages('pip install', packageNames)
  }

  async installSystemPackages(packageNames: string | string[]) {
    await this.installPackages('sudo apt-get install -y', packageNames)
  }

  private async installPackages(command: string, packageNames: string | string[]) {
    if (Array.isArray(packageNames)) {
      packageNames = packageNames.join(' ')
    }

    packageNames = packageNames.trim()
    if (packageNames.length === 0) {
      return
    }

    const out = await this.process.startAndWait(`${command} ${packageNames}`)

    if (out.exitCode !== 0) {
      throw new Error(`Failed to install package ${packageNames}: ${out.stderr}`)
    }
  }
}
